File: JavaScript/Automata.js
const crypto = require("crypto");
const { performance } = require("perf_hooks");

console.time("Tiempo de ejecucion");

class Nodo {
    constructor(partida, cuerpo) {
        this.partida = partida;
        this.cuerpo = cuerpo;
        this.firmaDigital = this.calcularFirma();
    }

    calcularFirma() {
        const contenido = `${this.partida} ${this.cuerpo.join(" ")}`;
        return crypto.createHash("sha256").update(contenido).digest("hex");
    }
}

function generarLista(n, k) {
    let listaNodos = [];

    let tiempoActual = new Date().toLocaleString("es-VE", { timeZone: "America/Caracas"});
    let primeraFirma = crypto.createHash("sha256").update(`${tiempoActual}`).digest("hex");
    let contenidoInicial = `${tiempoActual} ${primeraFirma}`;

    for(let i=0; i<n; i++){
        let partida = (i == 0)? contenidoInicial : listaNodos[i-1].firmaDigital;
        let cuerpo = Array.from({ length: k}, () => Math.floor(Math.random() * 100000) + 1);
        let nodo = new Nodo(partida, cuerpo);
        listaNodos.push(nodo);
    }
    return listaNodos;
}


const nValues = [3, 10, 200];
const kValues = [4, 200, 10];
const tiemposEjecucion = [];


nValues.forEach((n, index) => {
    let k = kValues[index];

    let inicio = performance.now();  
    let nodos = generarLista(n, k);
    let fin = performance.now(); 

    let tiempo = (fin - inicio).toFixed(5);
    console.log(`Tiempo para n=${n}, k=${k} = ${tiempo}`);
    tiemposEjecucion.push(`Tiempo ${String.fromCharCode(97 + index)}: ${tiempo} ms`);

    console.log(`\n=== Prueba con n=${n}, k=${k} ===`);
    nodos.forEach((nodo, i) => {
        console.log(`Nodo ${i + 1}:`);
        console.log(`  Partida: ${nodo.partida}`);
        console.log(`  Cuerpo: ${nodo.cuerpo}`);
        console.log(`  Firma Digital: ${nodo.firmaDigital}`);
        console.log("-".repeat(40));
    });
});

console.log("\n=== Resumen de tiempos ===");
tiemposEjecucion.forEach(tiempo => console.log(tiempo));

File: Java/Automata.java
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;



class Nodo {
    String partida;
    List<Integer> cuerpo;
    String firmaDigital;

    public Nodo(String partida, List<Integer> cuerpo) {
        this.partida = partida;
        this.cuerpo = cuerpo;
        this.firmaDigital = calcularFirma();
    }

    private String calcularFirma() {
        StringBuilder contenido = new StringBuilder(partida);
        for (int num : cuerpo) {
            contenido.append(" ").append(num);
        }
        return hashSHA256(contenido.toString());
    }

    private String hashSHA256(String input) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(input.getBytes());
            StringBuilder hexString = new StringBuilder();
            for (byte b : hash) {
                hexString.append(String.format("%02x", b));
            }
            return hexString.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }
}

public class Automata {
    public static List<Nodo> generarLista(int n, int k) {
        List<Nodo> listaNodos = new ArrayList<>();
        Random random = new Random();

        
        String tiempoActual = LocalDateTime.now().format(DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm:ss"));
        String primeraFirma = new Nodo(tiempoActual, new ArrayList<>()).firmaDigital;
        primeraFirma = tiempoActual + " " + primeraFirma;
        
        for (int i = 0; i < n; i++) {
            String partida = (i == 0) ? primeraFirma : listaNodos.get(i - 1).firmaDigital;
            List<Integer> cuerpo = new ArrayList<>();
            for (int j = 0; j < k; j++) {
                cuerpo.add(random.nextInt(100000) + 1);
            }
            listaNodos.add(new Nodo(partida, cuerpo));
        }

        return listaNodos;
    }

    public static void main(String[] args) {
        int[] nValues = {3, 10, 200};
        int[] kValues = {4, 200, 10};
        double[] tiemposEjecucion = new double[nValues.length];

        for (int i = 0; i < nValues.length; i++) {
            int n = nValues[i];
            int k = kValues[i];

            long inicio = System.nanoTime();
            List<Nodo> nodos = generarLista(n, k);
            long fin = System.nanoTime();

            tiemposEjecucion[i] = (fin - inicio) / 1e6;  // Convertir a milisegundos

            System.out.println("\n=== Prueba con n=" + n + ", k=" + k + " ===");
            System.out.printf("Tiempo de ejecución: %.5f ms\n", tiemposEjecucion[i]);

            for (int j = 0; j < nodos.size(); j++) {
                System.out.println("Nodo " + (j + 1) + ":");
                System.out.println("  Partida: " + nodos.get(j).partida);
                System.out.println("  Cuerpo: " + nodos.get(j).cuerpo);
                System.out.println("  Firma Digital: " + nodos.get(j).firmaDigital);
                System.out.println("-".repeat(40));
            }
        }

        // Imprimir resumen de tiempos al final
        System.out.println("\n=== Resumen de tiempos ===");
        for (int i = 0; i < tiemposEjecucion.length; i++) {
            System.out.printf("Tiempo %c: %.5f ms\n", 'a' + i, tiemposEjecucion[i]);
        }
    }
}

